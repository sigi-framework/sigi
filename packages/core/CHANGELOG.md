# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [2.13.2](https://github.com/sigi-framework/sigi/compare/v2.13.1...v2.13.2) (2025-06-20)


### Bug Fixes

* **core:** hmr in vite ([23dcbdc](https://github.com/sigi-framework/sigi/commit/23dcbdc0df2b791a5fbbf2450f01de472d54e173))





## [2.13.1](https://github.com/sigi-framework/sigi/compare/v2.13.0...v2.13.1) (2025-06-20)

**Note:** Version bump only for package @sigi/core





# [2.13.0](https://github.com/sigi-framework/sigi/compare/v2.12.2...v2.13.0) (2024-04-22)


### Features

* upgrade dependencies ([fe81b78](https://github.com/sigi-framework/sigi/commit/fe81b785aefabdda40ecc920290c569c4d29a851))





## [2.12.1](https://github.com/sigi-framework/sigi/compare/v2.12.0...v2.12.1) (2021-09-18)

### Bug Fixes

- **react:** state selector may use old closures ([69b19ce](https://github.com/sigi-framework/sigi/commit/69b19ced8f5da2802edf8c1dc793300af448b19e))

## [2.11.3](https://github.com/sigi-framework/sigi/compare/v2.11.2...v2.11.3) (2021-08-30)

**Note:** Version bump only for package @sigi/core

## [2.11.2](https://github.com/sigi-framework/sigi/compare/v2.11.1...v2.11.2) (2021-08-24)

### Bug Fixes

- **core:** skipCount should less equal than the filtered index ([ff38eb6](https://github.com/sigi-framework/sigi/commit/ff38eb6667cc8f0274a893211669925cd3e7f334))

## [2.10.8](https://github.com/sigi-framework/sigi/compare/v2.10.7...v2.10.8) (2021-06-16)

### Bug Fixes

- **ssr:** skipped actions should always be retried on client ([4000e27](https://github.com/sigi-framework/sigi/commit/4000e27bc1ca2ccc3f8e5a8440ce3384b218d8e1))

## [2.10.7](https://github.com/sigi-framework/sigi/compare/v2.10.6...v2.10.7) (2021-06-15)

### Bug Fixes

- **core:** skip from SSR logic ([c3e1ee8](https://github.com/sigi-framework/sigi/commit/c3e1ee8ee97732294c7bf9f3759dd4d7e284e041))

## [2.10.4](https://github.com/sigi-framework/sigi/compare/v2.10.3...v2.10.4) (2021-06-09)

### Bug Fixes

- **core:** compatible with es class property ([3f0e38a](https://github.com/sigi-framework/sigi/commit/3f0e38a8531f72ec62f50552c9dd866945245665))

## [2.10.3](https://github.com/sigi-framework/sigi/compare/v2.10.2...v2.10.3) (2021-06-09)

**Note:** Version bump only for package @sigi/core

## [2.10.2](https://github.com/sigi-framework/sigi/compare/v2.10.1...v2.10.2) (2021-05-26)

**Note:** Version bump only for package @sigi/core

# [2.10.0](https://github.com/sigi-framework/sigi/compare/v2.9.3...v2.10.0) (2021-05-13)

### Features

- **ssr:** force all epic to complete after disposed ([d89786a](https://github.com/sigi-framework/sigi/commit/d89786a4ead1ce84901543274a54f6ea6fa03e9b))

# [2.9.0](https://github.com/sigi-framework/sigi/compare/v2.8.6...v2.9.0) (2021-05-11)

### Bug Fixes

- dependencies list ([4a44cdd](https://github.com/sigi-framework/sigi/commit/4a44cdd65b1fa83ba949f6993c2708d2f6c2b722))

### Features

- **core:** ssr effect now could be marked retry on client ([116df59](https://github.com/sigi-framework/sigi/commit/116df5952c8f45cda94a0d352a5c538b78307ffe))

### Performance Improvements

- **core:** reduce js engine hidden path mismatch ([3224344](https://github.com/sigi-framework/sigi/commit/32243447ba72b7aa957cc52ef7460aaef118e175))
- **react:** fast code path on server ([f7df497](https://github.com/sigi-framework/sigi/commit/f7df49708808f5421f40798151108280ab669359))
- **ssr:** add fast path for ssr di logic ([8cd2f5a](https://github.com/sigi-framework/sigi/commit/8cd2f5a4517fed2f01058cc39d410af63fa53a3b))

## [2.8.6](https://github.com/sigi-framework/sigi/compare/v2.8.5...v2.8.6) (2021-05-06)

**Note:** Version bump only for package @sigi/core

## [2.8.5](https://github.com/sigi-framework/sigi/compare/v2.8.4...v2.8.5) (2021-04-28)

**Note:** Version bump only for package @sigi/core

## [2.8.2](https://github.com/sigi-framework/sigi/compare/v2.8.1...v2.8.2) (2021-03-29)

**Note:** Version bump only for package @sigi/core

# [2.8.0](https://github.com/sigi-framework/sigi/compare/v2.7.0...v2.8.0) (2021-02-22)

**Note:** Version bump only for package @sigi/core

## [2.5.1](https://github.com/sigi-framework/sigi/compare/v2.5.0...v2.5.1) (2021-01-22)

### Bug Fixes

- **core:** setup store immediately after set default state ([d5d2631](https://github.com/sigi-framework/sigi/commit/d5d2631898645a6eb938f7cb8ff5bacac8a9efe6))

## [2.4.5](https://github.com/sigi-framework/sigi/compare/v2.4.4...v2.4.5) (2020-12-14)

### Bug Fixes

- **core:** epic do not follow rxjs control ([e27ab82](https://github.com/sigi-framework/sigi/commit/e27ab824d9b99846654087b840c8b2cccea9af6c))

## [2.4.4](https://github.com/sigi-framework/sigi/compare/v2.4.3...v2.4.4) (2020-11-27)

**Note:** Version bump only for package @sigi/core

## [2.4.3](https://github.com/sigi-framework/sigi/compare/v2.4.2...v2.4.3) (2020-11-22)

**Note:** Version bump only for package @sigi/core

## [2.4.2](https://github.com/sigi-framework/sigi/compare/v2.4.1...v2.4.2) (2020-10-15)

**Note:** Version bump only for package @sigi/core

## [2.4.1](https://github.com/sigi-framework/sigi/compare/v2.4.0...v2.4.1) (2020-08-27)

**Note:** Version bump only for package @sigi/core

# [2.4.0](https://github.com/sigi-framework/sigi/compare/v2.3.1...v2.4.0) (2020-07-28)

### Features

- **core:** get rid of symbols ([6d7a848](https://github.com/sigi-framework/sigi/commit/6d7a84801e4ce2a7be731ad13e26e56cd0970135))
- **react:** expose common actions to dispatcher hooks ([20dd063](https://github.com/sigi-framework/sigi/commit/20dd0631d2501e4b912210fef1f2b69addadb612))

## [2.3.1](https://github.com/sigi-framework/sigi/compare/v2.3.0...v2.3.1) (2020-07-09)

**Note:** Version bump only for package @sigi/core

# [2.3.0](https://github.com/sigi-framework/sigi/compare/v2.2.0...v2.3.0) (2020-07-05)

**Note:** Version bump only for package @sigi/core

# [2.2.0](https://github.com/sigi-framework/sigi/compare/v2.1.0...v2.2.0) (2020-06-28)

### Bug Fixes

- **core:** dispatching action on non-effects module causes dead loop ([1ad1100](https://github.com/sigi-framework/sigi/commit/1ad1100158bd131d7366fef165d4acca30e6739b))

# [2.1.0](https://github.com/sigi-framework/sigi/compare/v2.0.2...v2.1.0) (2020-06-02)

**Note:** Version bump only for package @sigi/core
