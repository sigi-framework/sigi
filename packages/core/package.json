{
  "name": "@sigi/core",
  "version": "2.13.0",
  "description": "Sigi core library",
  "keywords": [
    "RxJS",
    "TypeScript",
    "Sigi"
  ],
  "author": "LongYinan <longyinan.brooklyn@bytedance.com>",
  "homepage": "https://sigi.how",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "esm/index.js",
  "esnext": "next/index.js",
  "types": "esm/index.d.ts",
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  },
  "sideEffects": false,
  "files": [
    "dist",
    "esm",
    "next",
    "es2015",
    "src"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sigi-framework/sigi.git"
  },
  "bugs": {
    "url": "https://github.com/sigi-framework/sigi/issues"
  },
  "dependencies": {
    "@sigi/di": "workspace:*",
    "@sigi/types": "workspace:*",
    "tslib": "^2.8.1"
  },
  "peerDependencies": {
    "immer": ">=9",
    "rxjs": ">=7"
  }
}
