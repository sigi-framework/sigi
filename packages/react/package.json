{
  "name": "@sigi/react",
  "version": "2.12.2",
  "description": "React bindings for sigi framework",
  "keywords": [
    "Hooks",
    "RxJS",
    "TypeScript",
    "React",
    "Sigi"
  ],
  "author": "LongYinan <lynweklm@gmail.com>",
  "homepage": "https://sigi.how",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "esm/index.js",
  "browser": "esm/index.browser.js",
  "esnext": "next/index.browser.js",
  "types": "esm/index.d.ts",
  "publishConfig": {
    "registry": "https://registry.npmjs.org/",
    "access": "public"
  },
  "sideEffects": false,
  "files": [
    "dist",
    "esm",
    "next",
    "es2015",
    "src"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sigi-framework/sigi.git"
  },
  "bugs": {
    "url": "https://github.com/sigi-framework/sigi/issues"
  },
  "dependencies": {
    "@sigi/core": "^2.12.1",
    "@sigi/di": "^2.11.3",
    "@sigi/ssr": "^2.12.1",
    "@sigi/types": "^2.11.3",
    "tslib": "^2.6.1"
  },
  "peerDependencies": {
    "react": ">=18",
    "rxjs": ">=7",
    "use-sync-external-store": ">=1"
  },
  "devDependencies": {
    "@types/use-sync-external-store": "^0.0.6",
    "use-sync-external-store": "^1.2.0"
  }
}
