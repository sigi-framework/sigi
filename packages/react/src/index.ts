import { EffectModule, ActionOfEffectModule } from '@sigi/core'
import { ConstructorOf } from '@sigi/types'

import { useServerInstance } from './injectable-context'

export type StateSelector<S, U> = {
  (state: S): U
}

export type StateSelectorConfig<S, U> = {
  selector: StateSelector<S, U>
}

const SERVER_DISPATCHERS = new Proxy(Object.create(null), {
  apply() {
    if (process.env.NODE_ENV !== 'production') {
      console.warn(new Error('Dispatch while calling react-dom/server functions is forbidden'))
    }
  },
})

export function useDispatchers<M extends EffectModule<S>, S = any>(_A: ConstructorOf<M>): ActionOfEffectModule<M, S> {
  return SERVER_DISPATCHERS
}

export function useModuleState<M extends EffectModule<any>>(
  A: ConstructorOf<M>,
): M extends EffectModule<infer State> ? State : never

export function useModuleState<M extends EffectModule<any>, U>(
  A: ConstructorOf<M>,
  config: M extends EffectModule<infer State> ? StateSelectorConfig<State, U> : never,
): M extends EffectModule<infer State>
  ? (typeof config)['selector'] extends StateSelector<State, infer NewState>
    ? NewState
    : never
  : never

export function useModuleState<M extends EffectModule<any>, U>(
  A: ConstructorOf<M>,
  config?: M extends EffectModule<infer S> ? StateSelectorConfig<S, U> : never,
) {
  const { store } = useServerInstance(A)
  return typeof config?.selector === 'function' ? config.selector(store.state) : store.state
}

export function useModule<M extends EffectModule<any>>(
  A: ConstructorOf<M>,
): M extends EffectModule<infer State> ? [State, ActionOfEffectModule<M, State>] : never

export function useModule<M extends EffectModule<any>, U>(
  A: ConstructorOf<M>,
  config: M extends EffectModule<infer State> ? StateSelectorConfig<State, U> : never,
): M extends EffectModule<infer State>
  ? (typeof config)['selector'] extends StateSelector<State, infer NewState>
    ? [NewState, ActionOfEffectModule<M, State>]
    : never
  : never

export function useModule<M extends EffectModule<S>, U, S>(A: ConstructorOf<M>, config?: StateSelectorConfig<S, U>) {
  const effectModule = useServerInstance(A)
  const { store } = effectModule
  const appState = typeof config?.selector === 'function' ? config.selector(store.state) : store.state

  return [appState, SERVER_DISPATCHERS]
}

export { SSRContext } from './ssr-context'
export * from './injectable-context'
