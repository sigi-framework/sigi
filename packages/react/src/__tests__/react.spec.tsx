import 'reflect-metadata'

import { Module, EffectModule, Effect, ImmerReducer, Action } from '@sigi/core'
import { Draft } from 'immer'
import { useEffect, useState } from 'react'
import { create, act, ReactTestRenderer } from 'react-test-renderer'
import { Observable, timer } from 'rxjs'
import { mergeMap, map, endWith, switchMap } from 'rxjs/operators'
import * as Sinon from 'sinon'

import { useModuleState, useModule } from '../index.browser'

interface CountState {
  count: number
  name: string
}

@Module('CountModel')
class CountModel extends EffectModule<CountState> {
  defaultState = { count: 0, name: '' }

  @ImmerReducer()
  setCount(state: Draft<CountState>, count: number) {
    state.count = count
  }

  @ImmerReducer()
  setName(state: Draft<CountState>, name: string) {
    state.name = name
  }

  @Effect({
    ssr: true,
  })
  getCount(payload$: Observable<void>): Observable<Action> {
    return payload$.pipe(
      mergeMap(() =>
        timer(20).pipe(
          map(() => this.getActions().setCount(1)),
          endWith(this.terminate()),
        ),
      ),
    )
  }

  @Effect({
    payloadGetter: (ctx: { url: string }, skip) => ctx.url || skip,
    skipFirstClientDispatch: false,
  })
  skippedEffect(payload$: Observable<string>): Observable<Action> {
    return payload$.pipe(
      switchMap((name) =>
        timer(20).pipe(
          map(() => this.getActions().setName(name)),
          endWith(this.terminate()),
        ),
      ),
    )
  }
}

describe('React components test', () => {
  const stub = Sinon.stub()
  const spy = Sinon.spy()
  const TestComponent = () => {
    const [localCount, setLocalCount] = useState(0)
    const state = useModuleState(CountModel, {
      selector: (state) => state.count + localCount,
      dependencies: [localCount],
    })
    spy(state)
    stub.callsFake(() => {
      setLocalCount(localCount + 1)
    })
    return <div onClick={stub}>{state}</div>
  }

  afterEach(() => {
    spy.resetHistory()
  })

  it('should render once while initial rendering', () => {
    create(<TestComponent />)
    expect(spy.callCount).toBe(1)
    expect(spy.getCall(0).calledWith(0)).toBe(true)
  })

  it('should render three times while change local state which in dependencies list', () => {
    const reactNode = create(<TestComponent />)
    act(() => stub())
    expect(spy.callCount).toBe(2)
    expect(spy.getCall(0).calledWith(0)).toBe(true)
    expect(spy.getCall(1).calledWith(1)).toBe(true)
    expect(reactNode).toMatchSnapshot()
  })
})

describe('Hooks', () => {
  const resetStub = Sinon.stub()
  const setCountStub = Sinon.stub()
  const spy = Sinon.spy()
  const TestComponent = () => {
    const [state, dispatcher] = useModule(CountModel, {
      selector: (state) => state.count,
      dependencies: [],
    })
    spy()
    useEffect(() => {
      setCountStub.callsFake(() => {
        dispatcher.setCount(10)
      })
      resetStub.callsFake(() => {
        dispatcher.reset()
      })
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    return <div>{state}</div>
  }
  let testWrapper!: ReactTestRenderer

  beforeEach(() => {
    act(() => {
      testWrapper = create(<TestComponent />)
    })
  })

  afterEach(() => {
    act(() => {
      resetStub()
      testWrapper.unmount()
    })
    spy.resetHistory()
    setCountStub.reset()
    resetStub.reset()
  })

  it('could reset state from dispatcher', () => {
    expect(testWrapper.root.findByType('div').children[0]).toBe('0')

    act(() => {
      setCountStub()
    })
    expect(testWrapper.root.findByType('div').children[0]).toBe('10')

    act(() => {
      resetStub()
    })
    expect(testWrapper.root.findByType('div').children[0]).toBe('0')
  })

  it('should not re-render if return state shallow equaled', () => {
    const fooSpy = Sinon.spy()
    const FooComponent = () => {
      const state = useModuleState(CountModel, {
        selector: () => ({ name: 'John Doe' }),
        dependencies: [],
      })
      fooSpy()
      return <div>{state.name}</div>
    }

    let fooWrapper!: ReactTestRenderer
    act(() => {
      fooWrapper = create(<FooComponent />)
    })

    expect(fooWrapper.root.findByType('div').children[0]).toBe('John Doe')
    act(() => {
      setCountStub()
    })
    expect(fooSpy.callCount).toBe(1)
    expect(spy.callCount).toBe(2)
    fooSpy.resetHistory()
  })

  it('should re-render if return state not pass custom equality function', () => {
    const fooSpy = Sinon.spy()
    const FooComponent = () => {
      const state = useModuleState(CountModel, {
        selector: () => ({ name: 'John Doe' }),
        dependencies: [],
        equalFn: (a, b) => a === b,
      })
      fooSpy()
      return <div>{state.name}</div>
    }

    let fooWrapper!: ReactTestRenderer
    act(() => {
      fooWrapper = create(<FooComponent />)
    })

    expect(fooWrapper.root.findByType('div').children[0]).toBe('John Doe')
    act(() => {
      setCountStub()
    })
    expect(fooSpy.callCount).toBe(2)
    expect(spy.callCount).toBe(2)
    fooSpy.resetHistory()
  })

  it('should run selector with new closure', () => {
    const setPlusCountStub = Sinon.stub()
    const FooComponent = () => {
      const [plusCount, setPlusCount] = useState(1)
      const plusOneCount = useModuleState(CountModel, {
        selector: (state) => plusCount + state.count,
        dependencies: [plusCount],
      })

      useEffect(() => {
        setPlusCountStub.callsFake(() => {
          setPlusCount(2)
        })
      }, [])

      return <div>{plusOneCount}</div>
    }

    let fooWrapper!: ReactTestRenderer
    act(() => {
      fooWrapper = create(<FooComponent />)
    })

    expect(fooWrapper.root.findByType('div').children[0]).toBe('1')
    act(() => {
      setPlusCountStub()
    })
    expect(fooWrapper.root.findByType('div').children[0]).toBe('2')

    act(() => {
      setCountStub()
    })
    expect(fooWrapper.root.findByType('div').children[0]).toBe('12')

    setPlusCountStub.reset()
  })

  it('Child <-> Parent scenario', () => {
    const name = 'tj'

    const ChildComponent = () => {
      const [state, dispatcher] = useModule(CountModel)
      useEffect(() => {
        dispatcher.setName(name)
      }, [dispatcher])

      return <span>{state.name}</span>
    }

    const FooComponent = () => {
      const state = useModuleState(CountModel)
      return (
        <div>
          {state.name}
          <ChildComponent />
        </div>
      )
    }

    let fooWrapper!: ReactTestRenderer
    act(() => {
      fooWrapper = create(<FooComponent />)
    })
    expect(fooWrapper.root.findByType('span').children[0]).toBe(name)
    expect(fooWrapper.root.findByType('div').children[0]).toBe(name)
  })
})
