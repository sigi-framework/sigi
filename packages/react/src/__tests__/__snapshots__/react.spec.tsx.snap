// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`React components test > should render once while initial rendering 1`] = `"<div><div>John Doe: 0</div></div>"`;
